engine/src/state_chain_observer/client.rs

```diff
@@ -491,23 +524,77 @@ impl<BaseRpcClient: base_rpc_api::BaseRpcApi + Send + Sync + 'static, SignedExtr
 
 					// Often `finalized_header` returns a significantly newer latest block than the
 					// stream returns so we move the stream forward to this block
-					{
-						let finalised_header_hash =
-							base_rpc_client.latest_finalized_block_hash().await?;
-						let finalised_header =
-							base_rpc_client.block_header(finalised_header_hash).await?;
-						if finalized_block_stream.cache().number < finalised_header.number {
-							let blocks_to_skip =
-								finalized_block_stream.cache().number + 1..=finalised_header.number;
-							for block_number in blocks_to_skip {
+					let finalized_hash = base_rpc_client.latest_finalized_block_hash().await?;
+					let finalized_header = base_rpc_client.block_header(finalized_hash).await?;
+
+					let base_rpc_client = base_rpc_client.clone();
+					let mut finalized_block_stream: Pin<
+						Box<
```

Comment:
If you write the branches in the other order, you will not need to put that explicit type here. You'll need to use an if instead of match, I think because I don't think you can put a wildcard as the first condition.

Reply:
"I think because I do not think"

- René Descartes

Reply:
Using an `if` then means I have to have two else's that return the same error which is a little ugly, that's why I did it this way